pkg/controller/common/expectations/deletions.go

```diff
@@ -0,0 +1,75 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package expectations
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+// ExpectedPodDeletions stores UID of Pods that we did delete, but whose deletion may not be
+// done yet, or not visible yet in the cache.
+// It allows making sure we're not working with an out-of-date list of Pods that includes
+// Pods which do not exist anymore.
+type ExpectedPodDeletions struct {
+	client       k8s.Client
+	podDeletions map[types.NamespacedName]types.UID
+}
+
+// NewExpectedPodDeletions returns an initialized ExpectedPodDeletions.
+func NewExpectedPodDeletions(client k8s.Client) *ExpectedPodDeletions {
+	return &ExpectedPodDeletions{
+		client:       client,
+		podDeletions: make(map[types.NamespacedName]types.UID),
+	}
+}
+
+// ExpectDeletion registers an expected deletion for the given Pod.
+func (e *ExpectedPodDeletions) ExpectDeletion(pod corev1.Pod) {
+	e.podDeletions[k8s.ExtractNamespacedName(&pod)] = pod.UID
+}
+
+// CancelExpectedDeletion removes an expected deletion for the given Pod.
+// It's mostly used if a deletion call failed, but the expectation was already registered.
+func (e *ExpectedPodDeletions) CancelExpectedDeletion(pod corev1.Pod) {
```

Comment:
I think you're right. It's a leftover from the time when our deletions expectations where handled in another thread, at the watch events level. So we had to make sure we registered the expectation before doing the actual deletion (otherwise we could observe it before it's registered hence miss it).
Good catch 👍